src/box/lua/config/applier/credentials.lua

```diff
@@ -277,85 +282,291 @@ local function privileges_add_defaults(name, role_or_user, intermediate)
+    local credentials = configdata:get('credentials')
+
+    -- If credentials section in config is empty, skip applier.
+    if credentials == nil then
+        return {}
+    end
+
+    -- Tarantool has the following roles and users present by default on every
+    -- instance:
+    --
+    -- Default roles:
+    -- * super
+    -- * public
+    -- * replication
+    --
+    -- Default users:
+    -- * guest
+    -- * admin
+    --
+    -- These roles and users have according privileges pre-granted by design.
+    -- Credentials applier adds such privileges with `priviliges_add_default()`
+    -- when syncing. So, for the excessive (non-default) privs to be removed,
+    -- these roles and users must be present inside configuration at least in a
+    -- form of an empty table. Otherwise, the privileges will be left unchanged,
+    -- similar to all used-defined roles and users.
+
+    credentials.roles = credentials.roles or {}
+    credentials.roles['super'] = credentials.roles['super'] or {}
+    credentials.roles['public'] = credentials.roles['public'] or {}
+    credentials.roles['replication'] = credentials.roles['replication'] or {}
+
+    credentials.users = credentials.users or {}
+    credentials.users['guest'] = credentials.users['guest'] or {}
+    credentials.users['admin'] = credentials.users['admin'] or {}
+
+    return credentials
+end
+
+-- }}} Sync helpers
+
+-- {{{ Triggers for grants after obj creation
+
+-- This is fiber channel that contains all scheduled tasks
+-- in form of a table with the following format:
+-- {type = 'BLOCKING_FULL_SYNC'/'BACKGROUND_FULL_SYNC'/obj_type,
+--  name = obj_name}
+local sync_tasks
 
-    local grant_f = function(name, privs, obj_type, obj_name)
-        privs = table.concat(privs, ',')
-        log.debug('credentials.apply: ' .. role_or_user ..
-                  '.grant(%q, %q, %q, %q)', name, privs, obj_type, obj_name)
-        box.schema[role_or_user].grant(name, privs, obj_type, obj_name)
+-- The table contains flags for objects that should be synced in the trigger:
+-- [obj_type][obj_name] = true/nil
+local on_creation_flags = {
+    ['space'] = {},
+    ['function'] = {},
+    ['sequence'] = {},
+}
+
+-- Info about objects is stored inside system spaces
+-- box.space._space/_func/_sequence. Translate these names
+-- to obj_type values of space/function/sequence.
+local function translate_obj_type(obj_type)
+    if obj_type == '_space' then
+        return 'space'
+    elseif obj_type == '_func' then
+        return 'function'
+    elseif obj_type == '_sequence' then
+        return 'sequence'
     end
-    local revoke_f = function(name, privs, obj_type, obj_name)
-        privs = table.concat(privs, ',')
-        log.debug('credentials.apply: ' .. role_or_user ..
-                  '.revoke(%q, %q, %q, %q)', name, privs, obj_type, obj_name)
-        box.schema[role_or_user].revoke(name, privs, obj_type, obj_name)
+end
+
+local sync_privileges
```

Comment:
The function is not used till its definition, so we can just use `local function sync_privileges(<...>) <...> end` as before. IOW, there is no need to use a forward declaration.

Reply:
Thanks, missed that. Removed.